[WIP] Use System.Memory Span api for base58

hanabi1224 opened the pull request:
This PR tries to use Span api in Base58 decode method.  The benchmark below shows consistent perf gain

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-7600U CPU 2.80GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=2.2.300
  [Host]     : .NET Core 2.2.5 (CoreCLR 4.6.27617.05, CoreFX 4.6.27618.01), 64bit RyuJIT
  Job-WLOEEC : .NET Core 2.2.5 (CoreCLR 4.6.27617.05, CoreFX 4.6.27618.01), 64bit RyuJIT

Runtime=Core  Server=True  

```
|                      Method |                   text |          Mean |       Error |        StdDev |        Median | Rank |    Gen 0 | Gen 1 | Gen 2 |  Allocated |
|---------------------------- |----------------------- |--------------:|------------:|--------------:|--------------:|-----:|---------:|------:|------:|-----------:|
|            **Encode_Base58Btc** | **5VLCA(...)AL3rP [2154]** | **23,929.828 us** | **818.1862 us** | **2,412.4403 us** | **24,172.491 us** |   **16** | **656.2500** |     **-** |     **-** | **16073896 B** |
|          Encode_Base58BtcV2 | 5VLCA(...)AL3rP [2154] | 18,012.818 us | 222.0683 us |   185.4371 us | 18,096.431 us |   15 |        - |     - |     - |     8096 B |
| Encode_Base58Btc_SimpleBase | 5VLCA(...)AL3rP [2154] |  8,375.942 us | 163.4444 us |   167.8453 us |  8,350.053 us |   14 |        - |     - |     - |    11096 B |
|            Decode_Base58Btc | 5VLCA(...)AL3rP [2154] |  2,286.929 us | 122.9767 us |   360.6694 us |  2,222.245 us |   11 | 294.9219 |     - |     - |  7058304 B |
|          Decode_Base58BtcV2 | 5VLCA(...)AL3rP [2154] |  3,058.658 us |  71.6978 us |   136.4125 us |  3,013.892 us |   12 |        - |     - |     - |     1608 B |
| Decode_Base58Btc_SimpleBase | 5VLCA(...)AL3rP [2154] |  5,591.871 us | 111.5083 us |   295.7041 us |  5,488.316 us |   13 |        - |     - |     - |     3216 B |
|            **Encode_Base58Btc** |   **UT5UZ(...)UEhxL [78]** |     **61.175 us** |   **3.1170 us** |     **9.0430 us** |     **59.552 us** |   **10** |   **1.5869** |     **-** |     **-** |    **36920 B** |
|          Encode_Base58BtcV2 |   UT5UZ(...)UEhxL [78] |     25.507 us |   0.5077 us |     1.3813 us |     24.933 us |    9 |        - |     - |     - |      344 B |
| Encode_Base58Btc_SimpleBase |   UT5UZ(...)UEhxL [78] |     12.010 us |   0.2406 us |     0.2363 us |     11.981 us |    7 |   0.0153 |     - |     - |      480 B |
|            Decode_Base58Btc |   UT5UZ(...)UEhxL [78] |     15.881 us |   0.5342 us |     0.4461 us |     15.743 us |    8 |   0.8240 |     - |     - |    18928 B |
|          Decode_Base58BtcV2 |   UT5UZ(...)UEhxL [78] |      4.839 us |   0.0855 us |     0.0758 us |      4.832 us |    4 |        - |     - |     - |       88 B |
| Decode_Base58Btc_SimpleBase |   UT5UZ(...)UEhxL [78] |      6.885 us |   0.1307 us |     0.1556 us |      6.863 us |    6 |   0.0076 |     - |     - |      176 B |
|            **Encode_Base58Btc** |   **zUXE7(...)SEBPe [35]** |     **15.689 us** |   **0.3098 us** |     **0.2898 us** |     **15.657 us** |    **8** |   **0.4883** |     **-** |     **-** |    **11536 B** |
|          Encode_Base58BtcV2 |   zUXE7(...)SEBPe [35] |      5.027 us |   0.0736 us |     0.0653 us |      5.023 us |    5 |   0.0076 |     - |     - |      192 B |
| Encode_Base58Btc_SimpleBase |   zUXE7(...)SEBPe [35] |      2.511 us |   0.0429 us |     0.0381 us |      2.499 us |    3 |   0.0114 |     - |     - |      272 B |
|            Decode_Base58Btc |   zUXE7(...)SEBPe [35] |      6.903 us |   0.1213 us |     0.1013 us |      6.912 us |    6 |   0.2899 |     - |     - |     6520 B |
|          Decode_Base58BtcV2 |   zUXE7(...)SEBPe [35] |      1.018 us |   0.0191 us |     0.0204 us |      1.021 us |    1 |   0.0019 |     - |     - |       56 B |
| Decode_Base58Btc_SimpleBase |   zUXE7(...)SEBPe [35] |      1.380 us |   0.0270 us |     0.0239 us |      1.371 us |    2 |   0.0038 |     - |     - |      112 B |

hanabi1224 committed:
Use System.Memory Span api for base58

hanabi1224 commented on src/Multiformats.Base/Multibase.cs:
```diff
@@ -6,43 +6,39 @@ namespace Multiformats.Base
-        public static Multibase Base10 => _bases[MultibaseEncoding.Base10];
-        public static Multibase Base16 => _bases[MultibaseEncoding.Base16Lower];
-        public static Multibase Base32 => _bases[MultibaseEncoding.Base32Lower];
-        public static Multibase Base58 => _bases[MultibaseEncoding.Base58Btc];
```
Since it's singleton, there seems to be no reason to perform hashtable lookup on each property getter call

hanabi1224 commented:
Looks like CI is broken, let me fix it with another PR

tabrath commented:
I really like the improvements you've done here! The CI configuration used here is pretty old, no surprise it crashes now. Supporting net4.* is causing a lot of trouble. Maybe we should just bump the major version and drop net4 and only support netstandard?

hanabi1224 commented:
@tabrath I have created another PR to fix CI, it's ur call whether to drop stale framework target : )

hanabi1224 committed:
More optimization

hanabi1224 committed:
Encode with span api

tabrath commented:
Cool. I've been implementing spans for all base encodings in the last couple of days, we might want to coordinate our work. I'll make a new branch for that, then you can rebase onto it, I'll let you know when it's ready.

hanabi1224 committed:
Change decoding algo

hanabi1224 committed:
Merge remote-tracking branch 'origin/master' into span_api_base58

# Conflicts:
#	src/Multiformats.Base/Multiformats.Base.csproj
#	test/Multiformats.Base.Tests/Multiformats.Base.Tests.csproj

hanabi1224 commented:
@tabrath  I have updated the benchmark above, btw, do you have any concern that I switched the encode / decode algorithm by refering to unsafe code implementation here: https://github.com/ssg/SimpleBase/blob/master/src/Base58.cs

If you look at  memory allocation column, the GC overhead should be trivial in the new implementation

hanabi1224 committed:
Replace string builder with span

tabrath commented:
@hanabi1224 I got no concerns regarding the implementation, as long as it's correct and not a copy-paste of someone elses code - if it gains perf and/or reduced allocations, I'm more than happy with that.

hanabi1224 commented:
@tabrath Awesome! I can wait for your span PR to complete and then merge with that, sounds good?

rvagg commented:
closing due to inactivity, archiving repo
